Unclear which functions are not interpreted

When running `cn test ...` on various OpenSUT code, I often get the following error:

```
root@ac94c45769b6:/work/components/mission_protection_system/src# cn test --output-dir=test -I include --include=include/wars.h --magic-comment-char-dollar components/instrumentation_common.c
include/common.h:34:24: warning: experimental keyword 'cn_function' (use of experimental features is discouraged)
uint8_t c_NINSTR() /*$ cn_function NINSTR; $*/ { return NINSTR; }
                       ^~~~~~~~~~~ 
...
cn: internal error, uncaught exception:
    Failure("Uninterpreted CN functions not supported at runtime. Please provide a concrete function definition")
    Raised at Stdlib.failwith in file "stdlib.ml", line 29, characters 17-33
    Called from Cn__Cn_internal_to_ail.cn_to_ail_function_internal in file "backend/cn/lib/cn_internal_to_ail.ml", lines 2952-2954, characters 6-29
    Called from Stdlib__List.map in file "list.ml", line 86, characters 15-19
    Called from Cn__Executable_spec_internal.generate_c_functions_internal in file "backend/cn/lib/executable_spec_internal.ml", lines 376-382, characters 4-24
    Called from Cn__Executable_spec.main in file "backend/cn/lib/executable_spec.ml", line 233, characters 4-63
    Called from Dune__exe__Main.run_tests.(fun) in file "backend/cn/bin/main.ml", lines 495-504, characters 10-26
    Called from Cerb_colour.without_colour in file "util/cerb_colour.ml", line 40, characters 10-13
    Called from Dune__exe__Main.run_tests.(fun) in file "backend/cn/bin/main.ml", lines 476-531, characters 6-10
    Called from Dune__exe__Main.with_well_formedness_check in file "backend/cn/bin/main.ml", lines 157-168, characters 6-48
    Re-raised at Dune__exe__Main.with_well_formedness_check in file "backend/cn/bin/main.ml", line 176, characters 4-69
    Called from Cmdliner_term.app.(fun) in file "cmdliner_term.ml", line 24, characters 19-24
    Called from Cmdliner_eval.run_parser in file "cmdliner_eval.ml", line 35, characters 37-44
```

The source code hides certain functions/features to make `cn verify` pass, but it is difficult for me to understand *which* functions are causing the problems for `cn test`. I tried removing all modifications, but I am still hitting the same error. Is it possible to add a more informative output, ideally pointing at the offending line in the C source?

--- a/backend/cn/lib/cn_internal_to_ail.ml
+++ b/backend/cn/lib/cn_internal_to_ail.ml
@@ -2949,9 +2949,20 @@ let cn_to_ail_function_internal
       in
       (bs, Some (List.map mk_stmt ss))
     | Uninterp ->
-      failwith
-        "Uninterpreted CN functions not supported at runtime. Please provide a concrete \
-         function definition"
+      Cerb_colour.with_colour
+        (fun () ->
+          print_endline
+            Pp.(
+              plain
+                (Pp.item
+                   "Uninterpreted CN functions not supported at runtime. Please provide \
+                    a concrete function definition for"
+                   (space
+                    ^^^ squotes (Definition.Function.pp_sig (Sym.pp fn_sym) lf_def)
+                    ^^^ !^"at"
+                    ^^^ Locations.pp lf_def.loc))))
+        ();
+      exit 2
   in
   let ail_record_opt = generate_record_opt fn_sym lf_def.return_bt in
   let params = List.map (fun (sym, bt) -> (sym, bt_to_ail_ctype bt)) lf_def.args in

--- a/backend/cn/lib/definition.ml
+++ b/backend/cn/lib/definition.ml
@@ -33,18 +33,27 @@ module Function = struct
 
 
   let pp_args xs =
-    Pp.flow_map
-      (Pp.break 1)
-      (fun (sym, typ) -> Pp.parens (Pp.typ (Sym.pp sym) (BaseTypes.pp typ)))
-      xs
+    let doc =
+      Pp.flow_map
+        (Pp.break 1)
+        (fun (sym, typ) -> Pp.parens (Pp.typ (Sym.pp sym) (BaseTypes.pp typ)))
+        xs
+    in
+    if PPrint.requirement doc = 0 then
+      Pp.parens Pp.empty
+    else
+      doc
+
+
+  let pp_sig nm def =
+    let open Pp in
+    nm ^^ pp_args def.args ^^^ colon ^^^ BaseTypes.pp def.return_bt
 
 
   let pp nm def =
     let open Pp in
-    nm
-    ^^ colon
-    ^^^ pp_args def.args
-    ^^ colon
+    pp_sig nm def
+    ^^^ equals
     ^/^
     match def.body with
     | Uninterp -> !^"uninterpreted"

--- a/backend/cn/lib/definition.mli
+++ b/backend/cn/lib/definition.mli
@@ -20,6 +20,8 @@ module Function : sig
 
   val pp_args : (Cerb_frontend.Symbol.sym * unit BaseTypes.t_gen) list -> Pp.document
 
+  val pp_sig : Pp.document -> t -> Pp.document
+
   val pp : Pp.document -> t -> Pp.document
 
   val open_ : (Sym.t * 'a) list -> IndexTerms.t -> IndexTerms.t list -> IndexTerms.t

--- a/util/cerb_colour.ml
+++ b/util/cerb_colour.ml
@@ -34,6 +34,13 @@ let int_fg = function
 let do_colour =
   ref (Unix.isatty Unix.stdout)
 
+let with_colour f x =
+  let col = ! do_colour in
+  do_colour := true;
+  let r = f x in
+  do_colour := col;
+  r
+
 let without_colour f x =
   let col = ! do_colour in
   do_colour := false;

--- a/util/cerb_colour.mli
+++ b/util/cerb_colour.mli
@@ -16,6 +16,7 @@ type ansi_style =
 type ansi_format = ansi_style list
 
 val do_colour: bool ref
+val with_colour: ('a -> 'b) -> 'a -> 'b
 val without_colour: ('a -> 'b) -> 'a -> 'b
 val ansi_format: ?err:bool -> ansi_format -> string -> string
 val pp_ansi_format: ?err:bool -> ansi_format -> (unit -> PPrint.document) -> PPrint.document

